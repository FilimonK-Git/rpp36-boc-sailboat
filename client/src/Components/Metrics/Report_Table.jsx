import React from "react";
import Tasks from "./Tasks.jsx";

class ReportTable extends React.Component {
  constructor(props) {
    super(props);
  }

  sortTable(n) {
    var table,
      rows,
      switching,
      i,
      x,
      y,
      shouldSwitch,
      dir,
      switchcount = 0;
    table = document.getElementById("reportTable");
    switching = true;
    dir = "asc";
    while (switching) {
      switching = false;
      rows = table.rows;
      for (i = 1; i < rows.length - 1; i++) {
        shouldSwitch = false;
        x = rows[i].childNodes[n];
        y = rows[i + 1].childNodes[n];
        if (dir == "asc") {
          if (x.innerHTML.toLowerCase() > y.innerHTML.toLowerCase()) {
            shouldSwitch = true;
            break;
          }
        } else if (dir == "desc") {
          if (x.innerHTML.toLowerCase() < y.innerHTML.toLowerCase()) {
            shouldSwitch = true;
            break;
          }
        }
      }
      if (shouldSwitch) {
        rows[i].parentNode.insertBefore(rows[i + 1], rows[i]);
        switching = true;
        switchcount++;
      } else {
        if (switchcount == 0 && dir == "asc") {
          dir = "desc";
          switching = true;
        }
      }
    }
  }

  render() {
    return (
      <table id="reportTable">
        <thead>
          <tr>
            <th
              style={{ fontWeight: "bold" }}
              onClick={(e) => this.sortTable(0)}
            >
              Task
            </th>
            <th
              style={{ fontWeight: "bold" }}
              onClick={(e) => this.sortTable(1)}
            >
              Category
            </th>
            <th
              style={{ fontWeight: "bold" }}
              onClick={(e) => this.sortTable(2)}
            >
              Duration
            </th>
          </tr>
          {this.props.data.map((task, index) => (
            <Tasks
              task={task}
              key={index}
              timeUpdated={this.props.timeUpdated}
            />
          ))}
        </thead>
      </table>
    );
  }
}

export default ReportTable;
