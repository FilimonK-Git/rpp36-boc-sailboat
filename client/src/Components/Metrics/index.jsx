import React from "react";
import ReactDOM from "react-dom/client";
import axios from "axios";
import html2canvas from "html2canvas";
import { jsPDF } from "jspdf";
import { ReportPieChart } from "./Report_PieChart.jsx";
import { ReportBarChart } from "./Report_BarChart.jsx";
import ReportTable from "./Report_Table.jsx";

class Metrics extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      allData: [],
      categoriesANDcolor: [],
      categories: ["All"],
      timeFrame: "Today",
      category: "All",
      totalTime: "0 min",
      customStartDate: "2022-01-01",
      customEndDate: "2023-12-30",
    };
  }

  componentDidMount() {
    axios
      .get("/completedTasks", {
        params: {
          timeRange: "Today",
          customRange: [this.state.customStartDate, this.state.customEndDate],
          catg: "All",
        },
      })
      .then((allCompletedToDos) => {
        let allData = allCompletedToDos.data.results;

        let categoriesANDcolor = [];
        let categories = ["All"];
        for (let cat of allCompletedToDos.data.results) {
          categoriesANDcolor.push([cat.category, cat.color]);
          if (!categories.includes(cat.category)) {
            categories.push(cat.category);
          }
        }

        let chartLegend = [];
        let chartColors = [];
        for (let cat of categoriesANDcolor) {
          if (!chartLegend.includes(cat[0])) {
            chartLegend.push(cat[0]);
            chartColors.push(cat[1]);
          }
        }

        let catgDurations = {};
        for (let key of categories) {
          if (key !== "All") {
            catgDurations[key] = [];
          }
        }

        for (let key in catgDurations) {
          for (let i = 0; i < allData.length; i++) {
            if (key === allData[i].category) {
              catgDurations[key].push([
                allData[i].start_time,
                allData[i].end_time,
              ]);
            }
          }
        }

        let totalTimeSpent = 0;
        let chartData = [];
        for (let catg in catgDurations) {
          let catgTotalSec = 0;

          for (let i = 0; i < catgDurations[catg].length; i++) {
            let start = new Date(catgDurations[catg][i][0]);
            let end = new Date(catgDurations[catg][i][1]);

            let start_sec = start.getTime() / 1000;
            let end_sec = end.getTime() / 1000;
            let difference = Math.abs(start_sec - end_sec);
            catgTotalSec += difference;
          }

          catgDurations[catg].totalSec = catgTotalSec;
          chartData.push(catgTotalSec);
          catgDurations[catg].totalDuration = this.secondsToHms(catgTotalSec);
          totalTimeSpent += catgTotalSec;
        }
        catgDurations.totalTimeSpent = totalTimeSpent;
        let totalTime = this.secondsToHms(catgDurations.totalTimeSpent);

        let sortedCatg = [this.state.category];
        for (let catg of categories) {
          if (catg !== this.state.category) {
            sortedCatg.push(catg);
          }
        }

        this.setState({
          categoriesANDcolor,
          categories: sortedCatg,
          allData,
          totalTime,
        });
      })
      .catch((err) => {
        console.error("componentDidMount ERR", err);
      });
  }

  specifyCategory(input, timeR) {
    axios
      .get("/completedTasksPerCatg", {
        params: {
          timeRange: timeR,
          customRange: [this.state.customStartDate, this.state.customEndDate],
          catg: input,
        },
      })
      .then((allCatgData) => {
        let allData = allCatgData.data.results;

        let categoriesANDcolor = [];
        let categories = ["All"];
        for (let cat of allCatgData.data.results) {
          categoriesANDcolor.push([cat.category, cat.color]);
          if (!categories.includes(cat.category)) {
            categories.push(cat.category);
          }
        }

        let chartLegend = [];
        let chartColors = [];
        for (let cat of categoriesANDcolor) {
          if (!chartLegend.includes(cat[0])) {
            chartLegend.push(cat[0]);
            chartColors.push(cat[1]);
          }
        }

        let catgDurations = {};
        for (let key of categories) {
          if (key !== "All") {
            catgDurations[key] = [];
          }
        }

        for (let key in catgDurations) {
          for (let i = 0; i < allData.length; i++) {
            if (key === allData[i].category) {
              catgDurations[key].push([
                allData[i].start_time,
                allData[i].end_time,
              ]);
            }
          }
        }

        let totalTimeSpent = 0;
        let chartData = [];
        for (let catg in catgDurations) {
          let catgTotalSec = 0;

          for (let i = 0; i < catgDurations[catg].length; i++) {
            let start = new Date(catgDurations[catg][i][0]);
            let end = new Date(catgDurations[catg][i][1]);

            let start_sec = start.getTime() / 1000;
            let end_sec = end.getTime() / 1000;
            let difference = Math.abs(start_sec - end_sec);
            catgTotalSec += difference;
          }

          catgDurations[catg].totalSec = catgTotalSec;
          chartData.push(catgTotalSec);
          catgDurations[catg].totalDuration = this.secondsToHms(catgTotalSec);
          totalTimeSpent += catgTotalSec;
        }
        catgDurations.totalTimeSpent = totalTimeSpent;
        let totalTime = this.secondsToHms(catgDurations.totalTimeSpent);

        this.setState({
          allData,
          categoriesANDcolor,
          category: input,
          timeFrame: timeR,
          totalTime,
        });
      })
      .catch((err) => {
        console.error("specifyCategory ERR", err);
      });
  }

  secondsToHms(d) {
    d = Number(d);
    let h = Math.floor(d / 3600);
    let m = Math.floor((d % 3600) / 60);
    let s = Math.floor((d % 3600) % 60);
    let hDisplay = h > 0 ? h + (h == 1 ? " hour " : " hours ") : "";
    let mDisplay = m > 0 ? m + (m == 1 ? " minute " : " minutes ") : "";
    let sDisplay = s > 0 ? s + (s == 1 ? " second" : " seconds") : "";
    return hDisplay + mDisplay + sDisplay;
  }

  specifyTimeframe(input, catg) {
    if (input === "Custom") {
      this.setState({ timeFrame: input });
    } else {
      axios
        .get("/completedTasks", {
          params: {
            timeRange: input,
            customRange: [this.state.customStartDate, this.state.customEndDate],
            catg: catg,
          },
        })
        .then((allCompletedToDos) => {
          let allData = allCompletedToDos.data.results;
          let categoriesANDcolor = [];
          let categories = ["All"];
          for (let cat of allCompletedToDos.data.results) {
            categoriesANDcolor.push([cat.category, cat.color]);
            if (!categories.includes(cat.category)) {
              categories.push(cat.category);
            }
          }

          let chartLegend = [];
          let chartColors = [];
          for (let cat of categoriesANDcolor) {
            if (!chartLegend.includes(cat[0])) {
              chartLegend.push(cat[0]);
              chartColors.push(cat[1]);
            }
          }

          let catgDurations = {};
          for (let key of categories) {
            if (key !== "All") {
              catgDurations[key] = [];
            }
          }

          for (let key in catgDurations) {
            for (let i = 0; i < allData.length; i++) {
              if (key === allData[i].category) {
                catgDurations[key].push([
                  allData[i].start_time,
                  allData[i].end_time,
                ]);
              }
            }
          }

          let totalTimeSpent = 0;
          let chartData = [];
          for (let catg in catgDurations) {
            let catgTotalSec = 0;

            for (let i = 0; i < catgDurations[catg].length; i++) {
              let start = new Date(catgDurations[catg][i][0]);
              let end = new Date(catgDurations[catg][i][1]);

              let start_sec = start.getTime() / 1000;
              let end_sec = end.getTime() / 1000;
              let difference = Math.abs(start_sec - end_sec);
              catgTotalSec += difference;
            }

            catgDurations[catg].totalSec = catgTotalSec;
            chartData.push(catgTotalSec);
            catgDurations[catg].totalDuration = this.secondsToHms(catgTotalSec);
            totalTimeSpent += catgTotalSec;
          }
          catgDurations.totalTimeSpent = totalTimeSpent;
          let totalTime = this.secondsToHms(catgDurations.totalTimeSpent);

          this.setState({
            categoriesANDcolor,
            categories,
            allData,
            timeFrame: input,
            totalTime,
          });
        })
        .catch((err) => {
          console.error("specifyTimeframe ERR", err);
        });
    }
  }

  timeUpdated(input) {
    if (this.state.category === "All") {
      axios
        .get("/completedTasks", {
          params: {
            timeRange: this.state.timeFrame,
            customRange: [this.state.customStartDate, this.state.customEndDate],
            catg: "All",
          },
        })
        .then((allCompletedToDos) => {
          let allData = allCompletedToDos.data.results;
          let categoriesANDcolor = [];
          let categories = ["All"];
          for (let cat of allCompletedToDos.data.results) {
            categoriesANDcolor.push([cat.category, cat.color]);
            if (!categories.includes(cat.category)) {
              categories.push(cat.category);
            }
          }

          let chartLegend = [];
          let chartColors = [];
          for (let cat of categoriesANDcolor) {
            if (!chartLegend.includes(cat[0])) {
              chartLegend.push(cat[0]);
              chartColors.push(cat[1]);
            }
          }

          let catgDurations = {};
          for (let key of categories) {
            if (key !== "All") {
              catgDurations[key] = [];
            }
          }

          for (let key in catgDurations) {
            for (let i = 0; i < allData.length; i++) {
              if (key === allData[i].category) {
                catgDurations[key].push([
                  allData[i].start_time,
                  allData[i].end_time,
                ]);
              }
            }
          }

          let totalTimeSpent = 0;
          let chartData = [];
          for (let catg in catgDurations) {
            let catgTotalSec = 0;

            for (let i = 0; i < catgDurations[catg].length; i++) {
              let start = new Date(catgDurations[catg][i][0]);
              let end = new Date(catgDurations[catg][i][1]);

              let start_sec = start.getTime() / 1000;
              let end_sec = end.getTime() / 1000;
              let difference = Math.abs(start_sec - end_sec);
              catgTotalSec += difference;
            }

            catgDurations[catg].totalSec = catgTotalSec;
            chartData.push(catgTotalSec);
            catgDurations[catg].totalDuration = this.secondsToHms(catgTotalSec);
            totalTimeSpent += catgTotalSec;
          }
          catgDurations.totalTimeSpent = totalTimeSpent;
          let totalTime = this.secondsToHms(catgDurations.totalTimeSpent);

          this.setState({ categoriesANDcolor, categories, allData, totalTime });
        })
        .catch((err) => {
          console.error("timeUpdated ERR", err);
        });
    } else {
      this.specifyCategory(input, this.state.timeFrame);
    }
  }

  customStartDate(input) {
    this.setState({ customStartDate: input });
  }

  customEndDateAndSearch(customEndDate, catg) {
    axios
      .get("/completedTasks", {
        params: {
          timeRange: "Custom",
          customRange: [this.state.customStartDate, customEndDate],
          catg: catg,
        },
      })
      .then((allCompletedToDos) => {
        let allData = allCompletedToDos.data.results;
        let categoriesANDcolor = [];
        let categories = ["All"];
        for (let cat of allCompletedToDos.data.results) {
          categoriesANDcolor.push([cat.category, cat.color]);
          if (!categories.includes(cat.category)) {
            categories.push(cat.category);
          }
        }

        let chartLegend = [];
        let chartColors = [];
        for (let cat of categoriesANDcolor) {
          if (!chartLegend.includes(cat[0])) {
            chartLegend.push(cat[0]);
            chartColors.push(cat[1]);
          }
        }

        let catgDurations = {};
        for (let key of categories) {
          if (key !== "All") {
            catgDurations[key] = [];
          }
        }

        for (let key in catgDurations) {
          for (let i = 0; i < allData.length; i++) {
            if (key === allData[i].category) {
              catgDurations[key].push([
                allData[i].start_time,
                allData[i].end_time,
              ]);
            }
          }
        }

        let totalTimeSpent = 0;
        let chartData = [];
        for (let catg in catgDurations) {
          let catgTotalSec = 0;

          for (let i = 0; i < catgDurations[catg].length; i++) {
            let start = new Date(catgDurations[catg][i][0]);
            let end = new Date(catgDurations[catg][i][1]);

            let start_sec = start.getTime() / 1000;
            let end_sec = end.getTime() / 1000;
            let difference = Math.abs(start_sec - end_sec);
            catgTotalSec += difference;
          }

          catgDurations[catg].totalSec = catgTotalSec;
          chartData.push(catgTotalSec);
          catgDurations[catg].totalDuration = this.secondsToHms(catgTotalSec);
          totalTimeSpent += catgTotalSec;
        }
        catgDurations.totalTimeSpent = totalTimeSpent;
        let totalTime = this.secondsToHms(catgDurations.totalTimeSpent);

        this.setState({
          categoriesANDcolor,
          categories,
          allData,
          timeFrame: "Custom",
          category: catg,
          totalTime,
        });
      })
      .catch((err) => {
        console.error("customEndDateAndSearch ERR", err);
      });
  }

  printDocument() {
    const printable = document.getElementById("Print");

    html2canvas(printable, {
      useCORS: true,
      allowTaint: true,
      scrollY: -window.scrollY,
    }).then((canvas) => {
      const image = canvas.toDataURL("image/jpeg", 1.0);
      const doc = new jsPDF("p", "px", "a4");
      const pageWidth = doc.internal.pageSize.getWidth();
      const pageHeight = doc.internal.pageSize.getHeight();

      const widthRatio = pageWidth / canvas.width;
      const heightRatio = pageHeight / canvas.height;
      const ratio = widthRatio > heightRatio ? heightRatio : widthRatio;

      const canvasWidth = canvas.width * ratio;
      const canvasHeight = canvas.height * ratio;

      const marginX = (pageWidth - canvasWidth) / 2;
      const marginY = (pageHeight - canvasHeight) / 2;

      doc.addImage(image, "JPEG", marginX, marginY, canvasWidth, canvasHeight);

      doc.save(`${this.state.timeFrame}_report.pdf`);
    });
  }

  render() {
    return (
      <div id="Print">
        <br></br>
        <hr></hr>
        <div className="taskDescription">
          {this.state.totalTime ? (
            <div>
              You have <em>{this.state.allData.length}</em> tasks completed{" "}
              <em>{this.state.timeFrame}</em>. <br></br>You spent a total time
              of <b>{this.state.totalTime}</b>{" "}
            </div>
          ) : (
            <div>
              You have no tasks completed <em>{this.state.timeFrame}</em>.
            </div>
          )}
        </div>

        <hr></hr>
        <br></br>

        <div className="dropDownContainer">
          <div>
            <label htmlFor="timeframe">Time frame:</label>
            <select
              name="timeframe"
              id="timeframe"
              onChange={(e) => {
                this.specifyTimeframe(e.target.value, this.state.category);
              }}
            >
              <option value="Today">Today</option>
              <option value="This Week">This Week</option>
              <option value="This Month">This Month</option>
              <option value="Custom">Custom</option>
            </select>
          </div>
          {this.state.timeFrame === "Custom" ? (
            <div>
              <div>
                <label htmlFor="customFrom">From: </label>
                <input
                  type="date"
                  id="customFrom"
                  name="customFrom"
                  onChange={(e) => {
                    this.customStartDate(e.target.value);
                  }}
                ></input>
              </div>
              <div>
                <label htmlFor="customTo">To: </label>
                <input
                  type="date"
                  id="customTo"
                  name="customTo"
                  onChange={(e) => {
                    this.customEndDateAndSearch(
                      e.target.value,
                      this.state.category
                    );
                  }}
                ></input>
              </div>
            </div>
          ) : (
            ""
          )}
          <br></br>
          <div>
            <label htmlFor="Category">Category:</label>
            <select
              name="Category"
              id="Category"
              onChange={(e) => {
                this.specifyCategory(e.target.value, this.state.timeFrame);
              }}
            >
              {this.state.categories.map((cat, i) => {
                return (
                  <option value={`${cat}`} key={i}>
                    {cat}
                  </option>
                );
              })}
            </select>
          </div>
        </div>
        <br></br>

        {this.state.allData.length === 0 ? (
          <div>
            <img
              className="zeroTaskImg"
              src="https://media1.tenor.com/images/e38013fbff0156a3f2f5aab705af70e1/tenor.gif?itemid=26990487"
            />
          </div>
        ) : (
          <div>
            <div>
              <div>
                {this.state.category === "All" ? (
                  <ReportPieChart data={this.state} />
                ) : (
                  <ReportBarChart data={this.state} />
                )}
              </div>
              <br></br>
              <ReportTable
                data={this.state.allData}
                timeUpdated={this.timeUpdated.bind(this)}
              />
            </div>
            <br></br>
            <div className="reportPDFbtnDIV">
              <button
                className="reportPDFbtn"
                onClick={this.printDocument.bind(this)}
              >
                Download Report (PDF)
              </button>
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default Metrics;
