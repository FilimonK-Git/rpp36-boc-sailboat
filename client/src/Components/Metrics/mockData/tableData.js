module.exports.mockData = [
  {
    todo_id: 5,
    user_id: 1,
    task: "Buy water",
    description: "Shop",
    category_id: 2,
    start_time: "2022-10-26T14:00:00.000Z",
    end_time: "2022-10-26T14:30:00.000Z",
    completed: true,
    appointment: false,
    category: "Shop",
    color: "#86dff2",
  },
  {
    todo_id: 1,
    user_id: 1,
    task: "Morning debrief",
    description: "Meeting",
    category_id: 1,
    start_time: "2022-10-26T11:00:00.000Z",
    end_time: "2022-10-26T12:30:00.000Z",
    completed: true,
    appointment: false,
    category: "Meeting",
    color: "#d6e9f2",
  },
  {
    todo_id: 12,
    user_id: 1,
    task: "AM gym",
    description: "Workout",
    category_id: 4,
    start_time: "2022-10-26T11:00:00.000Z",
    end_time: "2022-10-26T11:45:00.000Z",
    completed: true,
    appointment: false,
    category: "Workout",
    color: "#8539b8",
  },
  {
    todo_id: 8,
    user_id: 1,
    task: "trash out",
    description: "Misc",
    category_id: 3,
    start_time: "2022-10-26T16:00:00.000Z",
    end_time: "2022-10-26T16:10:00.000Z",
    completed: true,
    appointment: false,
    category: "Misc",
    color: "#7ee9cc",
  },
  {
    todo_id: 13,
    user_id: 1,
    task: "PM gym",
    description: "Workout",
    category_id: 4,
    start_time: "2022-10-26T15:00:00.000Z",
    end_time: "2022-10-26T15:15:00.000Z",
    completed: true,
    appointment: false,
    category: "Workout",
    color: "#8539b8",
  },
];

module.exports.mockState = {
  allData: [
    {
      todo_id: 5,
      user_id: 1,
      task: "Buy water",
      description: "Shop",
      category_id: 2,
      start_time: "2022-10-26T14:00:00.000Z",
      end_time: "2022-10-26T14:30:00.000Z",
      completed: true,
      appointment: false,
      category: "Shop",
      color: "#86dff2",
    },
    {
      todo_id: 1,
      user_id: 1,
      task: "Morning debrief",
      description: "Meeting",
      category_id: 1,
      start_time: "2022-10-26T11:00:00.000Z",
      end_time: "2022-10-26T12:30:00.000Z",
      completed: true,
      appointment: false,
      category: "Meeting",
      color: "#d6e9f2",
    },
    {
      todo_id: 12,
      user_id: 1,
      task: "AM gym",
      description: "Workout",
      category_id: 4,
      start_time: "2022-10-26T11:00:00.000Z",
      end_time: "2022-10-26T11:45:00.000Z",
      completed: true,
      appointment: false,
      category: "Workout",
      color: "#8539b8",
    },
    {
      todo_id: 8,
      user_id: 1,
      task: "trash out",
      description: "Misc",
      category_id: 3,
      start_time: "2022-10-26T16:00:00.000Z",
      end_time: "2022-10-26T16:10:00.000Z",
      completed: true,
      appointment: false,
      category: "Misc",
      color: "#7ee9cc",
    },
    {
      todo_id: 13,
      user_id: 1,
      task: "PM gym",
      description: "Workout",
      category_id: 4,
      start_time: "2022-10-26T15:00:00.000Z",
      end_time: "2022-10-26T15:15:00.000Z",
      completed: true,
      appointment: false,
      category: "Workout",
      color: "#8539b8",
    },
  ],
  categoriesANDcolor: [
    ["Shop", "#86dff2"],
    ["Meeting", "#d6e9f2"],
    ["Workout", "#8539b8"],
    ["Misc", "#7ee9cc"],
    ["Workout", "#8539b8"],
  ],
  categories: ["All", "Shop", "Meeting", "Workout", "Misc"],
  timeFrame: "Today",
  category: "All",
  totalTime: "30 min",
  customStartDate: "2022-01-01",
  customEndDate: "2023-12-30",
};
