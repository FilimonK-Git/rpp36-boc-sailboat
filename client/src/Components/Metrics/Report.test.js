import React from "react";
import ReactDOM from "react-dom/client";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { act } from "react-dom/test-utils";
import userEvent from "@testing-library/user-event";
import { jest } from "@jest/globals";
import { useState, useEffect } from "react";
import "jest-canvas-mock";
import $ from "jquery";
import Metrics from "./index.jsx";
import ReportTable from "./Report_Table.jsx";
import Tasks from "./Tasks.jsx";
import { ReportPieChart } from "./Report_PieChart.jsx";
import { ReportBarChart } from "./Report_BarChart.jsx";
import { mockData, mockState } from "./mockData/tableData.js";
import renderer from "react-test-renderer";
import { Chart as ChartJS, ArcElement, Tooltip, Legend } from "chart.js";
import { Pie, Doughnut } from "react-chartjs-2";
import axios from "axios";

ChartJS.register(ArcElement, Tooltip, Legend);

describe("Report Metric Test Suite", () => {
  describe("Testing Report Table Rendering", () => {
    it("should have three table headers and test data information", () => {
      render(<ReportTable data={mockData} />);
      expect(screen.getByText("Task")).toBeInTheDocument();
      expect(screen.getByText("Category")).toBeInTheDocument();
      expect(screen.getByText("Duration")).toBeInTheDocument();
      expect(screen.getByText("Meeting")).toBeInTheDocument();
      expect(screen.getByText("Shop")).toBeInTheDocument();
      expect(screen.getByText("Buy water")).toBeInTheDocument();
      expect(screen.getByText("Morning debrief")).toBeInTheDocument();
      expect(screen.getByText("PM gym")).toBeInTheDocument();
    });
    it("should sort table alphabetically by task name", () => {
      const { container } = render(<ReportTable data={mockData} />);
      fireEvent.mouseOver(container.querySelector("th"));

      const menuIcon = container.querySelector("th");

      var unsortedTable = $("#reportTable tr")
        .map(function (i, v) {
          var $td = $("td", this);
          return {
            id: ++i,
            task: $td.eq(0).text(),
            category: $td.eq(1).text(),
            duration: $td.eq(2).text(),
          };
        })
        .get();

      //first click
      fireEvent.click(menuIcon);

      var sortedASCByTaskName = $("#reportTable tr")
        .map(function (i, v) {
          var $td = $("td", this);
          return {
            id: ++i,
            task: $td.eq(0).text(),
            category: $td.eq(1).text(),
            duration: $td.eq(2).text(),
          };
        })
        .get();

      expect(unsortedTable[1].task).toBe(" Buy water");
      expect(sortedASCByTaskName[1].task).toBe(" AM gym");

      //second click
      fireEvent.click(menuIcon);

      var sortedDESByTaskName = $("#reportTable tr")
        .map(function (i, v) {
          var $td = $("td", this);
          return {
            id: ++i,
            task: $td.eq(0).text(),
            category: $td.eq(1).text(),
            duration: $td.eq(2).text(),
          };
        })
        .get();

      // console.log("sorted", sortedDESByTaskName);
      expect(sortedDESByTaskName[1].task).toBe(" trash out");
    });
  });

  describe("Testing dropdown menus", () => {
    it("should have 3 options for time frame dropdown selection", () => {
      render(<Metrics data={mockState} />);
      var options = $("select").map(function (i, v) {
        var $option = $("option", this);
        return {
          id: ++i,
          firstOption: $option.eq(0).text(),
          secondOption: $option.eq(1).text(),
          thirdOption: $option.eq(2).text(),
          fourthOption: $option.eq(3).text(),
        };
      });

      expect(options[0].firstOption).toBe("Today");
      expect(options[0].secondOption).toBe("This Week");
      expect(options[0].thirdOption).toBe("This Month");
      expect(options[0].fourthOption).toBe("Custom");
    });
  });

  describe("Test creation of two charts with test data", () => {
    let state = {
      allData: mockData,
      categoriesANDcolor: [
        ["Shop", "#86dff2"],
        ["Meeting", "#d6e9f2"],
        ["Workout", "#8539b8"],
        ["Misc", "#7ee9cc"],
        ["Workout", "#8539b8"],
      ],
      categories: ["All", "Shop", "Meeting", "Workout", "Misc"],
      timeFrame: "Today",
      category: "All",
      totalTime: "30 min",
      customStartDate: "2022-01-01",
      customEndDate: "2023-12-30",
    };

    it("renders pie chart successfully", () => {
      const pieChart = renderer
        .create(<ReportPieChart data={state} />)
        .toJSON();
      expect(pieChart).toMatchSnapshot();

      const testRenderer = renderer.create(<ReportPieChart data={state} />);
      const testInstance = testRenderer.root;
      expect(testInstance.findByType(ReportPieChart).props.data).toEqual(state);
    });

    it("renders bar chart successfully", () => {
      const barChart = renderer.create(
        <ReportBarChart data={state} />
      ).children;
      expect(barChart).toMatchSnapshot();

      const testRenderer = renderer.create(<ReportBarChart data={state} />);
      const testInstance = testRenderer.root;
      expect(testInstance.findByType(ReportBarChart).props.data).toEqual(state);
    });
  });
});
