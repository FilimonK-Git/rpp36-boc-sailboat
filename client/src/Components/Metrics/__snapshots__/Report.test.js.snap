// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Metric Test Suite Test creation of two charts with test data renders bar chart successfully 1`] = `undefined`;

exports[`Report Metric Test Suite Test creation of two charts with test data renders pie chart successfully 1`] = `
<canvas
  className="charts"
  height={150}
  role="img"
  width={300}
/>
`;
